distsql, sql, ui: update SQL metrics

rjnn opened the pull request:
Update the SQL metrics to track DistSQL queries separately from
non-DistSQL queries in the Admin UI. Also fixed a TODO to stop
double-counting transactions that are retried. This is the first PR 
in a series to add all the features listed in #12143.

The Admin UI part is WIP, since we might want to list these under a new "SQL internals" page, since users may not know or care what DistSQL is. But the work in the backend is the same.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/12299)
<!-- Reviewable:end -->

arjunravinarayan commented:
cc @vivekmenezes @andreimatei

cuongdo commented:
Looks mostly good to me. @andreimatei should look at the changes to the txn retry code.

---

Review status: 0 of 10 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/metric_test.go, line 41 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12299#-KYt5cikZ6NGEnsI1_ls:-KYt5cilkqZAI2u8tMG8:b-rkkq13) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2feaaf6ddb7b3587b5d9f78022ac87505c88226/pkg/sql/metric_test.go#L41)):*
> ```Go
> 		txnBeginCount    int64
> 		selectCount      int64
> 		distSQLCount     int64
> ```

is there a test where `distSQLCount > 0`?

---

*[pkg/sql/metric_test.go, line 235 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12299#-KYt4SRXW7w8Z-XyR2_B:-KYt4SRXW7w8Z-XyR2_C:bv53l6i) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2feaaf6ddb7b3587b5d9f78022ac87505c88226/pkg/sql/metric_test.go#L235)):*
> ```Go
> 		t.Error(err)
> 	}
> 	if err := checkCounterEQ(s, sql.MetaSelect, 0); err != nil {
> ```

So, the semantics of the `SELECT` counter are changing? I'm not sure if this is the right thing. There's a  argument to be made that counters should log all attempts instead of just successful executions. Also, don't the counters for `UPDATE`s and other queries now have different semantics from `SELECT` with your change?

---

*[pkg/ui/app/containers/nodeGraphs.tsx, line 104 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12299#-KYszEm5zHiAW_tPRote:-KYszEm6ccu-PuaLtcIT:b-kc09ns) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2feaaf6ddb7b3587b5d9f78022ac87505c88226/pkg/ui/app/containers/nodeGraphs.tsx#L104)):*
> ```XML
>             <LineGraph title="Reads" sources={sources} tooltip={`The average number of SELECT statements per second ${specifier}.`}>
>               <Axis format={ d3.format(".1f") }>
>                 <Metric name="cr.node.sql.select.count" title="Non-distsql Reads" nonNegativeRate />
> ```

`distsql` is capitalized differently than the following line. Also, shouldn't these be `DistSQL`, because `SQL` is an acronym?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12299#-:-KYt5j4g647Ko35eIcYw:bs71ltg)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
I think it'd be nice if we could use different dimensions of a metric to count different related things. For example, ideally, distsql/classic would be one dimension of the statements metric. original/auto-retry/user-directed-retry would be another dimension. @mrtracy, our metrics don't support having multiple dimensions, do they?

---

Review status: 0 of 10 files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[pkg/internal/client/txn.go, line 482 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12299#-KYts3_1vpBahJTX65GM:-KYts3_1vpBahJTX65GN:b-jmk5ar) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2feaaf6ddb7b3587b5d9f78022ac87505c88226/pkg/internal/client/txn.go#L482)):*
> ```Go
> // used.
> func (txn *Txn) Exec(
> 	opt TxnExecOptions, fn func(txn *Txn, opt *TxnExecOptions, isRetry bool) error,
> ```

mmm do we really need to add this param to this core interface that just one client needs? That one client can capture a variable from the outer scope, just like it does for a bunch of other stuff.

---

*[pkg/sql/executor.go, line 199 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12299#-KYu-KzjcX91tyjm-SCb:-KYu-KzjcX91tyjm-SCc:bgbb1oz) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2feaaf6ddb7b3587b5d9f78022ac87505c88226/pkg/sql/executor.go#L199)):*
> ```Go
> 	// Transient stats.
> 	Latency            *metric.Histogram
> 	SelectCount        *metric.Counter
> ```

comment whether the two overlap or not. As I've said below, I think the first should be a super-set of the 2nd.

---

*[pkg/sql/executor.go, line 978 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12299#-KYturR9r7_u_jmEyG52:-KYturR9r7_u_jmEyG53:b-t4qoze) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2feaaf6ddb7b3587b5d9f78022ac87505c88226/pkg/sql/executor.go#L978)):*
> ```Go
> //  so that nested BEGIN statements are caught.
> // stmtTimestamp: Used as the statement_timestamp().
> // isRetry: A boolean that is set for retries so that we don't double count in
> ```

well so the way you wrote the code, `isRetry` is gonna be set for automatic retries, but not for user-directed retries. I think that this is probably the behavior we want wrt counting queries in metrics,  but this should be documented somewhere and this param should be named accordingly.

---

*[pkg/sql/executor.go, line 1340 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12299#-KYttKXZYPcV1GdCu5tM:-KYttKXZYPcV1GdCu5tN:b-vybmq3) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2feaaf6ddb7b3587b5d9f78022ac87505c88226/pkg/sql/executor.go#L1340)):*
> ```Go
> // updateStmtCounts updates metrics for the number of times the different types of SQL
> // statements have been received by this node.
> func (e *Executor) updateStmtCounts(stmt parser.Statement, isRetry bool) {
> ```

nit: i'd have the parent not call this on retries

---

*[pkg/sql/executor.go, line 1350 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12299#-KYtyJTMJtg2tMqsYs5M:-KYtyJTMJtg2tMqsYs5N:b-qqvbin) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2feaaf6ddb7b3587b5d9f78022ac87505c88226/pkg/sql/executor.go#L1350)):*
> ```Go
> 		e.TxnBeginCount.Inc(1)
> 	case *parser.Select:
> 		// Do nothing as SELECTs are counted separately, depending on whether
> ```

I think it's very unfortunate to start counting these things in different places. How about we count selects here just like everything else (and so a select would always be counted), and we just count the distsql metric separately.

---

*[pkg/sql/metric_test.go, line 115 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12299#-KYtzo0SyU30rBZC0iZ5:-KYtzo0SyU30rBZC0iZ6:b-jqqdpi) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2feaaf6ddb7b3587b5d9f78022ac87505c88226/pkg/sql/metric_test.go#L115)):*
> ```Go
> 		}
> 
> 		// Everything after this query will also fail, so quit now to avoid deluge of errors.
> ```

well if you're giving this test some love, how about changing it to the use sub-tests and checking the delta between current value and the value at the beginning on the sub-test, instead of doing this early fail?

---

*[pkg/sql/metric_test.go, line 184 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12299#-KYtzRTrSL4Ivc9trDsu:-KYtzRTs0GVnNGmV3CnH:b-ck0ki4) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2feaaf6ddb7b3587b5d9f78022ac87505c88226/pkg/sql/metric_test.go#L184)):*
> ```Go
> 		t.Fatal(err)
> 	}
> 	query := "INSERT INTO db.t VALUES ('key', 'marker')"
> ```

superfluous change? same below

---

*[pkg/sql/metric_test.go, line 235 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12299#-KYt4SRXW7w8Z-XyR2_B:-KYu-he7ZhtNiTn9O_iV:b-eigira) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2feaaf6ddb7b3587b5d9f78022ac87505c88226/pkg/sql/metric_test.go#L235)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary><blockquote>

So, the semantics of the `SELECT` counter are changing? I'm not sure if this is the right thing. There's a  argument to be made that counters should log all attempts instead of just successful executions. Also, don't the counters for `UPDATE`s and other queries now have different semantics from `SELECT` with your change?
</blockquote></details>

yeah, how did this change?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12299#-:-KYu03Pzm6Ws7qirCx0u:byf37xx)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented on pkg/internal/client/txn.go:
```diff
@@ -478,7 +478,9 @@ func (e *AutoCommitError) Error() string {
 // used.
-func (txn *Txn) Exec(opt TxnExecOptions, fn func(txn *Txn, opt *TxnExecOptions) error) (err error) {
+func (txn *Txn) Exec(
+	opt TxnExecOptions, fn func(txn *Txn, opt *TxnExecOptions, isRetry bool) error,
```
Can you make isRetry a part of the TxnExecOptions?

Commit:
distsql, sql, ui: update SQL metrics

Update the SQL metrics to track DistSQL queries separately from
non-DistSQL queries in the Admin UI. Also fixed a TODO to stop
double-counting transactions that are retried.

arjunravinarayan commented:
Thank you for the reviews! Apologies for waiting a day, my email client didn't show me that there were reviews, so I didn't notice until I manually checked.

@andreimatei I think you're right. I've refactored it that way. Now `distsql.select.count` is a subset of `select.count`, and the semantics of `select.count` are left unchanged.

---

Review status: 0 of 5 files reviewed at latest revision, 11 unresolved discussions, some commit checks pending.

---

*[pkg/sql/executor.go, line 199 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12299#-KYu-KzjcX91tyjm-SCb:-KYytT0ysGsWf3Q-2jU5:b-5d88r0) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2feaaf6ddb7b3587b5d9f78022ac87505c88226/pkg/sql/executor.go#L199)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

comment whether the two overlap or not. As I've said below, I think the first should be a super-set of the 2nd.
</blockquote></details>

Done. I agree with you that `DistSQLSelectCount` should be a subset of `SelectCount` and have refactored.

---

*[pkg/sql/executor.go, line 978 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12299#-KYturR9r7_u_jmEyG52:-KYz-FXMJ70RhZp3MJKV:b-eegna6) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2feaaf6ddb7b3587b5d9f78022ac87505c88226/pkg/sql/executor.go#L978)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

well so the way you wrote the code, `isRetry` is gonna be set for automatic retries, but not for user-directed retries. I think that this is probably the behavior we want wrt counting queries in metrics,  but this should be documented somewhere and this param should be named accordingly.
</blockquote></details>

renamed to isAutomaticRetry, and added a comment in line 598 (where it first is created).

---

*[pkg/sql/executor.go, line 1340 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12299#-KYttKXZYPcV1GdCu5tM:-KYyv1fGoL3-Zp3LN1mj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2feaaf6ddb7b3587b5d9f78022ac87505c88226/pkg/sql/executor.go#L1340)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: i'd have the parent not call this on retries
</blockquote></details>

Done.

---

*[pkg/sql/executor.go, line 1350 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12299#-KYtyJTMJtg2tMqsYs5M:-KYyv2t2spVq8P_c50KI:b6vn8uf) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2feaaf6ddb7b3587b5d9f78022ac87505c88226/pkg/sql/executor.go#L1350)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I think it's very unfortunate to start counting these things in different places. How about we count selects here just like everything else (and so a select would always be counted), and we just count the distsql metric separately.
</blockquote></details>

Done. You are correct, this is a much better way.

---

*[pkg/sql/metric_test.go, line 41 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12299#-KYt5cikZ6NGEnsI1_ls:-KYzXhzzALwd9ByIDCCj:b-12znsy) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2feaaf6ddb7b3587b5d9f78022ac87505c88226/pkg/sql/metric_test.go#L41)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary><blockquote>

is there a test where `distSQLCount > 0`?
</blockquote></details>

There is now! It's also renamed to `distSQLSelectCount`.

---

*[pkg/sql/metric_test.go, line 115 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12299#-KYtzo0SyU30rBZC0iZ5:-KYz1oUHrResW1WEwjoy:b-6g2uhc) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2feaaf6ddb7b3587b5d9f78022ac87505c88226/pkg/sql/metric_test.go#L115)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

well if you're giving this test some love, how about changing it to the use sub-tests and checking the delta between current value and the value at the beginning on the sub-test, instead of doing this early fail?
</blockquote></details>

Done. I've earned some cleanup karma for today. We use subtests, and the test values are now defined in terms of increments so its easier to understand whats going on. PTAL.

---

*[pkg/sql/metric_test.go, line 184 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12299#-KYtzRTrSL4Ivc9trDsu:-KYyweZnXdaOnLNrwP_x:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2feaaf6ddb7b3587b5d9f78022ac87505c88226/pkg/sql/metric_test.go#L184)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

superfluous change? same below
</blockquote></details>

Done.

---

*[pkg/sql/metric_test.go, line 235 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12299#-KYt4SRXW7w8Z-XyR2_B:-KYz-xihbb7tkF77IKNY:b5dsvzf) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2feaaf6ddb7b3587b5d9f78022ac87505c88226/pkg/sql/metric_test.go#L235)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

yeah, how did this change?
</blockquote></details>

This was a mistake. Reverted. The semantics of `SELECT` should not change.

---

*[pkg/ui/app/containers/nodeGraphs.tsx, line 104 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12299#-KYszEm5zHiAW_tPRote:-KYyvJTmc2z7mcyPt4aI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2feaaf6ddb7b3587b5d9f78022ac87505c88226/pkg/ui/app/containers/nodeGraphs.tsx#L104)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary><blockquote>

`distsql` is capitalized differently than the following line. Also, shouldn't these be `DistSQL`, because `SQL` is an acronym?
</blockquote></details>

Done.

---

*[pkg/internal/client/txn.go, line 482 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12299#-KYpDsYBbb0mxWKNDJqg-r1-482:-KYz-9d4FPcSPCyilWPz:b-odg5lf) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2feaaf6ddb7b3587b5d9f78022ac87505c88226/pkg/internal/client/txn.go#L482)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

Can you make isRetry a part of the TxnExecOptions?
</blockquote></details>

Followed Andrei's suggestion below instead.

---

*[pkg/internal/client/txn.go, line 482 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12299#-KYts3_1vpBahJTX65GM:-KYz-DIW80PMoRj7DVi4:bncpi6t) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2feaaf6ddb7b3587b5d9f78022ac87505c88226/pkg/internal/client/txn.go#L482)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

mmm do we really need to add this param to this core interface that just one client needs? That one client can capture a variable from the outer scope, just like it does for a bunch of other stuff.
</blockquote></details>

Yes, that's a better idea. Thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12299#-:-KYzZl7nSuC7_ivo5Zx3:b-g1hpoc)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
@andreimatei: what do you mean by dimensions? I'm unclear whether the latest changes capture your intent.

Do you mean that the new DistSQL select count should be subordinate to the select count? If so, are you thinking about something like `sql.select.distributed-count` (I don't love the name, but it's an example).

---

Review status: 0 of 5 files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12299#-:-KZ-FhiV0_qgjC_BUOvD:bqfxla4)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Reviewed 1 of 10 files at r1, 9 of 9 files at r2.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12299)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 4 of 5 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/sql/executor.go, line 1007 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12299#-KZ2PfBtHBm30NGvjAo0:-KZ2PfBtHBm30NGvjAo1:b23mqxa) ([raw file](https://github.com/cockroachdb/cockroach/blob/c681563b40223a29119567b75ce64d174306cffb/pkg/sql/executor.go#L1007)):*
> ```Go
> 
> 	// Do not double count automatically retried transactions.
> 	if !isAutomaticRetry {
> ```

nit: I would move calling `updateStmtCounts` to the parent and not add the retry param to this method. Unless you need to pass the argument to `execStmt()` as I was saying in a comment below.

---

*[pkg/sql/executor.go, line 1329 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12299#-KZ2Q1JGQ84V6-o32nVI:-KZ2Q1JGQ84V6-o32nVJ:b-zhyc3x) ([raw file](https://github.com/cockroachdb/cockroach/blob/c681563b40223a29119567b75ce64d174306cffb/pkg/sql/executor.go#L1329)):*
> ```Go
> 		switch stmt.(type) {
> 		case *parser.Select:
> 			e.DistSQLSelectCount.Inc(1)
> ```

don't you need to pipe the retry argument to this level, and increment the metric conditionally?

---

*[pkg/sql/metric_test.go, line 115 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12299#-KYtzo0SyU30rBZC0iZ5:-KZ2PDAFssbxdl42iCoe:b-1fux23) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2feaaf6ddb7b3587b5d9f78022ac87505c88226/pkg/sql/metric_test.go#L115)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

Done. I've earned some cleanup karma for today. We use subtests, and the test values are now defined in terms of increments so its easier to understand whats going on. PTAL.
</blockquote></details>

much better test now :)

---

*[pkg/sql/metric_test.go, line 52 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12299#-KZ2Ok4m-YFF31BBAH8h:-KZ2Ok4m-YFF31BBAH8i:b-7vmfg7) ([raw file](https://github.com/cockroachdb/cockroach/blob/c681563b40223a29119567b75ce64d174306cffb/pkg/sql/metric_test.go#L52)):*
> ```Go
> 
> 	var testcases = []queryCounter{
> 		// The counts are increments for each query.
> ```

nit: I think this comment, or the comment on queryCounter struct, should say something about deltas.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12299#-:-KZ2QPrhIfCVT7-SyZpX:bnfp4nl)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 4 of 5 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12299#-:-KZ2mEWUhnGfdM4Bonnq:bnfp4nl)*
<!-- Sent from Reviewable.io -->

arjunravinarayan commented:
TFTRs!

---

Review status: 2 of 5 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[pkg/sql/executor.go, line 1007 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12299#-KZ2PfBtHBm30NGvjAo0:-KZ2zRTb4eZIOeVrjO9p:bybmot9) ([raw file](https://github.com/cockroachdb/cockroach/blob/c681563b40223a29119567b75ce64d174306cffb/pkg/sql/executor.go#L1007)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: I would move calling `updateStmtCounts` to the parent and not add the retry param to this method. Unless you need to pass the argument to `execStmt()` as I was saying in a comment below.
</blockquote></details>

Doing the latter, so keeping it here.

---

*[pkg/sql/executor.go, line 1329 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12299#-KZ2Q1JGQ84V6-o32nVI:-KZ2z2llUzQ1_YbLf4Nr:bsywv9p) ([raw file](https://github.com/cockroachdb/cockroach/blob/c681563b40223a29119567b75ce64d174306cffb/pkg/sql/executor.go#L1329)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

don't you need to pipe the retry argument to this level, and increment the metric conditionally?
</blockquote></details>

Yes, you're right. Done.

---

*[pkg/sql/metric_test.go, line 52 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12299#-KZ2Ok4m-YFF31BBAH8h:-KZ2z14Nchwhy8d6YGrR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c681563b40223a29119567b75ce64d174306cffb/pkg/sql/metric_test.go#L52)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: I think this comment, or the comment on queryCounter struct, should say something about deltas.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12299#-:-KZ2V06smoe7O_RBnWO9:bz5mair)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 2 of 5 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/sql/executor.go, line 1329 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/12299#-KZ2Q1JGQ84V6-o32nVI:-Kf82bDbyPZDP_Uheo09:bskly89) ([raw file](https://github.com/cockroachdb/cockroach/blob/c681563b40223a29119567b75ce64d174306cffb/pkg/sql/executor.go#L1329)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

Yes, you're right. Done.
</blockquote></details>

I think there's an error here - the `execDistSQL()` should not be conditional on `isAutomaticRetry` - only the  metric.
Fixing in https://github.com/cockroachdb/cockroach/pull/14116

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12299)*
<!-- Sent from Reviewable.io -->
